pkg/kubelet/cm/dra/plugin/plugin.go

```diff
@@ -0,0 +1,175 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dra
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	utilversion "k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/klog/v2"
+)
+
+const (
+	// DRAPluginName is the name of the in-tree DRA Plugin.
+	DRAPluginName = "kubernetes.io/dra"
+)
+
+// draPlugins map keeps track of all registered DRA plugins on the node
+// and their corresponding sockets.
+var draPlugins = &PluginsStore{}
+
+// RegistrationHandler is the handler which is fed to the pluginwatcher API.
+type RegistrationHandler struct{}
+
+var PluginHandler = &RegistrationHandler{}
+
+// RegisterPlugin is called when a plugin can be registered.
+func (h *RegistrationHandler) RegisterPlugin(pluginName string, endpoint string, versions []string) error {
+	klog.InfoS("Register new DRA plugin", "name", pluginName, "endpoint", endpoint)
+
+	highestSupportedVersion, err := h.validateVersions("RegisterPlugin", pluginName, versions)
+	if err != nil {
+		return err
+	}
+
+	// Storing endpoint of newly registered DRA Plugin into the map, where plugin name will be the key
+	// all other DRA components will be able to get the actual socket of DRA plugins by its name.
+	draPlugins.Set(pluginName, &Plugin{
+		endpoint:                endpoint,
+		highestSupportedVersion: highestSupportedVersion,
+	})
+
+	return nil
+}
+
+// Return the highest supported version.
+func highestSupportedVersion(versions []string) (*utilversion.Version, error) {
+	if len(versions) == 0 {
+		return nil, errors.New(log("DRA driver reporting empty array for supported versions"))
+	}
+
+	var highestSupportedVersion *utilversion.Version
+
+	var theErr error
+
+	for i := len(versions) - 1; i >= 0; i-- {
+		currentHighestVer, err := utilversion.ParseGeneric(versions[i])
+		if err != nil {
+			theErr = err
+
+			continue
+		}
+
+		if currentHighestVer.Major() > 1 {
+			// DRA currently only has version 1.x
+			continue
+		}
+
+		if highestSupportedVersion == nil || highestSupportedVersion.LessThan(currentHighestVer) {
+			highestSupportedVersion = currentHighestVer
+		}
+	}
+
+	if highestSupportedVersion == nil {
+		return nil, fmt.Errorf(
+			"could not find a highest supported version from versions (%v) reported by this driver: %+v",
+			versions, theErr)
+	}
+
+	if highestSupportedVersion.Major() != 1 {
+		return nil, fmt.Errorf("highest supported version reported by driver is %v, must be v1.x", highestSupportedVersion)
+	}
+
+	return highestSupportedVersion, nil
+}
+
+func (h *RegistrationHandler) validateVersions(
+	callerName string,
+	pluginName string,
+	versions []string,
+) (*utilversion.Version, error) {
+	if len(versions) == 0 {
+		return nil, errors.New(
+			log(
+				"%s for DRA driver %q failed. Plugin returned an empty list for supported versions",
+				callerName,
+				pluginName,
+			),
+		)
+	}
+
+	// Validate version
+	newDriverHighestVersion, err := highestSupportedVersion(versions)
```

Comment:
we seem to mix the terms "plugin" and "driver" alot throughout this code, which one should it be?

Reply:
it should be 'plugin'. I'll replace 'driver' with 'plugin' in the code.